internal/controlplane/handlers_oauth.go

```diff
@@ -221,34 +263,110 @@ func (s *Server) processCallback(ctx context.Context, w http.ResponseWriter, r *
 	return nil
 }
 
-func (s *Server) generateOAuthToken(ctx context.Context, providerName string, code string,
-	stateData db.GetProjectIDBySessionStateRow) error {
-	// generate a new OAuth2 config for the given provider
-	oauthConfig, err := s.providerAuthFactory(providerName, true)
+func (s *Server) processAppCallback(ctx context.Context, w http.ResponseWriter, r *http.Request,
+	pathParams map[string]string) error {
+	state := r.URL.Query().Get("state")
+	code := r.URL.Query().Get("code")
+
+	// Configure tracing
+	span := trace.SpanFromContext(ctx)
+	span.SetName("server.HandleProviderCallback")
+	span.SetAttributes(attribute.Key("code").String(code))
+	defer span.End()
+
+	provider := pathParams["provider"]
+
+	// Telemetry logging
+	logger.BusinessRecord(ctx).Provider = provider
+
+	token, err := s.exchangeCodeForToken(ctx, provider, code)
 	if err != nil {
-		return fmt.Errorf("error creating OAuth config: %w", err)
-	}
-	if oauthConfig == nil {
-		return errors.New("oauth2.Config is nil")
+		return err
 	}
 
-	token, err := oauthConfig.Exchange(ctx, code)
+	installationID, err := strconv.ParseInt(r.URL.Query().Get("installation_id"), 10, 64)
 	if err != nil {
-		return fmt.Errorf("error exchanging code for token: %w", err)
+		return fmt.Errorf("unable to parse installation ID to integer: %v", err)
 	}
 
-	_, err = s.providers.CreateGitHubOAuthProvider(ctx, providerName, db.ProviderClassGithub, *token, stateData)
+	err = s.providers.ValidateGitHubInstallationId(ctx, token, installationID)
 	if err != nil {
-		if errors.Is(err, providers.ErrInvalidTokenIdentity) {
-			return newHttpError(http.StatusForbidden, "User token mismatch").SetContents(
-				"The provided login token was associated with a different GitHub user.")
+		return newHttpError(http.StatusForbidden, "User installation ID mismatch").SetContents(
+			"The GitHub user does not have access to the requested installation.")
+	}
+
+	if state != "" {
```

Comment:
nit: Let's move this block to its own function for readability.

Reply:
I gave it a try, but it ends up looking weird, because this part creates the provider and extracts the redirect URL and then redirects. So I had a function called `enrollProviderAndGetRedirectURL` which was really strange. I'm open to other suggestions, maybe breaking out a subsection of this block.

Reply:
That's a good point. Maybe in case 1, where `state == ""` we redirect to the Minder frontend, since that will be the next step for them if they want to actually use Minder?

Reply:
👍